### PR TITLE
refactor: Use RelicApp in Server

### LINKED ISSUE #4101: Refactor request handling to use Router class from relic package
## Description
The current server implementation uses manual path and HTTP verb dispatch. This should be refactored to use the `Router` class from the `relic` package for better maintainability and extensibility.

## Affected Files
- `packages/serverpod/lib/src/server/server.dart:190`

## Context
```dart
// TODO: Use Router instead of manual dispatch on path and verb
if (uri.path == '/') {
  // Perform health checks
  ...
}
```

## Suggested Implementation
- Use the `Router` class from the `relic` package (already a dependency)
- Implement route matching and dispatching logic using relic's Router API
- Migrate existing manual dispatch code to use Router
- Add support for route parameters and pattern matching as provided by relic
- Maintain backward compatibility during transition
- Add tests for route matching edge cases

## Impact
Code maintainability, extensibility for adding new routes, better error handling

## Related
- PR #4085

### PR DESCRIPTION
**Pull request by nielsenko**
## Summary

Use Relic's router for endpoint dispatch in api server.

Closes: #4101

## Changes

- Migrated server to implement `RouterInjectable` and use `RelicApp` for request routing
- Removed `EndpointResult` classes.
- Changed `_handleUriCall` to return `Response` directly instead of `EndpointResult`
- Moved all exception-to-response mapping to the `_reportException` middleware
- Made `EndpointDispatchException` sealed to enable exhaustive pattern matching
- Simplified `NotAuthorizedException` to store `AuthenticationFailureReason` directly
- Updated `canUserAccessEndpoint` to return `AuthenticationFailureReason?` instead of result wrapper
- Removed `sessionLogId` from error responses to avoid exposing internal database IDs
- Removed helper methods `_toResponse` and `_logMalformedCalls` (no longer needed)
- Updated all existing tests to reflect new return types

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else.
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

## Breaking changes

None. This is an internal refactoring that doesn't affect the public API.

## Release Notes

* **New Features**
  * Added router injection capability for streamlined middleware and route registration.

* **Refactor**
  * Simplified authentication error handling, moving from verbose result objects to concise reason-based responses.
  * Restructured request dispatch and exception handling for improved maintainability.
  * Enhanced server startup architecture with injected application management.

### CODE REVIEW

**Comment by nielsenko** [comment]
@coderabbit review

**Comment by nielsenko** [comment]
@coderabbit review

File: packages/serverpod/lib/src/server/server.dart [hunk 1]
```diff
@@ -155,255 +174,163 @@ class Server {
...
+      });
+
+      // early exit on Method.options
+      if (isOptions) return Response.ok(headers: headers);
+
+      final result = await next(req);
+      return switch (result) {
+        Response() => result.copyWith(headers: headers),
+        _ => result,
+      };
```
[the remark this answers is not in the document]
Comment (nielsenko) [reply packages/serverpod/lib/src/server/server.dart:276]: Actually `result.headers` will always be empty here (currently), but we should perhaps not assume that.
Fixed in ba1561a

[the remark this answers is not in the document]
Comment (nielsenko) [reply packages/serverpod/lib/src/server/server.dart:276]: Actually I think it would be even better to not overwrite any existing headers here. I have changed that in eba4895e

File: packages/serverpod/lib/src/server/server.dart [hunk 2]
```diff
@@ -155,255 +174,170 @@ class Server {
...
+      responseBuffer.writeln('SADNESS ${DateTime.now().toUtc()}');
+    }
+    for (var issue in issues) {
+      responseBuffer.writeln(issue);
     }
+
+    return Response(
+      allOk ? io.HttpStatus.ok : io.HttpStatus.serviceUnavailable,
+      body: Body.fromString(responseBuffer.toString()),
+      //headers: headers,
```
Comment (SandPod) [inline packages/serverpod/lib/src/server/server.dart]: Suggestion: Remove comment.

File: packages/serverpod/lib/src/server/server.dart [hunk 3]
```diff
@@ -155,255 +174,170 @@ class Server {
...
-            case Body():
-              value = Response.ok(body: value, headers: headers);
-              continue response;
-            response:
-            case Response():
-              return value;
+  Handler _headers(Handler next) {
+    return (req) async {
+      final isOptions = req.method == Method.options;
+      // TODO: Make httpResponseHeaders a Headers object from the get-go.
```
Comment (SandPod) [inline packages/serverpod/lib/src/server/server.dart]: Question: Is this a short lived followup? Otherwise, please remove the TODO comment and file an issue for it.

File: packages/serverpod/lib/src/server/server.dart [hunk 4]
```diff
@@ -155,255 +174,170 @@ class Server {
...
+    if (allOk) {
+      responseBuffer.writeln('OK ${DateTime.now().toUtc()}');
+    } else {
+      responseBuffer.writeln('SADNESS ${DateTime.now().toUtc()}');
+    }
+    for (var issue in issues) {
+      responseBuffer.writeln(issue);
     }
+
+    return Response(
```
Comment (SandPod) [inline packages/serverpod/lib/src/server/server.dart]: Suggestion(NIT): Switch to early return in this method to simplify the code?

File: packages/serverpod/lib/src/server/server.dart [hunk 5]
```diff
@@ -155,255 +174,170 @@ class Server {
...
-            SerializationManager.encodeForProtocol(value),
-            mimeType: MimeType.json,
-          ),
-          headers: headers,
-        );
-    }
+      });
+
+      // early exit on Method.options
+      if (isOptions) return Response.ok(headers: headers);
```
Comment (SandPod) [inline packages/serverpod/lib/src/server/server.dart:263]: Suggestion: Can we register this as a separate route only accepting the options method?

File: packages/serverpod/lib/src/server/server.dart [hunk 6]
```diff
@@ -155,255 +174,170 @@ class Server {
...
+                    }
+                  })),
+        _ => result,
+      };
+    };
+  }
+
+  FutureOr<Result> _cloudStorage(Request req) async {
+    final uri = req.requestedUri;
+    assert(uri.path == '/serverpod_cloud_storage');
```
Comment (SandPod) [inline packages/serverpod/lib/src/server/server.dart:282]: Question: Why is this assertion needed? We should control this in the registered path, right?

File: packages/serverpod/lib/src/server/server.dart [hunk 7]
```diff
@@ -155,255 +174,170 @@ class Server {
...
+    return await _handleUriCall(uri, '', req);
```
Comment (SandPod) [inline packages/serverpod/lib/src/server/server.dart]: Discussion: Should we rename `_handleUriCall` to `_handleEndpointCall`?  Since it now only handles endpoints, right?

File: packages/serverpod/lib/src/server/server.dart [hunk 8]
```diff
@@ -540,9 +459,37 @@ class Server {
           methodCallContext.arguments,
         );
 
-        return ResultSuccess(
-          result,
-          sendAsRaw: methodCallContext.endpoint.sendAsRaw,
+        // Handle success - construct Response directly
+        var value = result;
```
Comment (SandPod) [inline packages/serverpod/lib/src/server/server.dart]: Comment: Not a big fan of this switch statement logic. I see that this was already here from before, so we might not need to prove it. But it would have been nice to break it up a bit.

File: packages/serverpod/lib/src/server/server.dart [hunk 9]
```diff
@@ -155,255 +174,170 @@ class Server {
...
-          this,
-          webSocket,
-          req,
-          () => _webSockets.remove(websocketKey),
-        );
+  ) {
+    return (req) async {
+      return WebSocketUpgrade((webSocket) async {
+        try {
+          // TODO(kasper): Should we keep doing this?
```
Comment (SandPod) [inline packages/serverpod/lib/src/server/server.dart]: Suggestion: Can we remove this comment now?

**Review by SandPod — COMMENTED** [review]
Changes looks good, have a couple of comments and minor suggestions.

File: packages/serverpod/lib/src/server/server.dart (hunk 3 above)
Comment (nielsenko) [reply packages/serverpod/lib/src/server/server.dart]: Issue is #4102 . Will remove.

File: packages/serverpod/lib/src/server/server.dart (hunk 5 above)
Comment (nielsenko) [reply packages/serverpod/lib/src/server/server.dart:263]: This is middleware. We don't have verb specific middleware routing in Relic currently.

File: packages/serverpod/lib/src/server/server.dart (hunk 6 above)
Comment (nielsenko) [reply packages/serverpod/lib/src/server/server.dart:282]: Yes, but `_handleUriCall` does use the uri. Right now it is basically handled like a regular generated endpoint.

File: packages/serverpod/lib/src/server/server.dart (hunk 8 above)
Comment (nielsenko) [reply packages/serverpod/lib/src/server/server.dart]: It builds up a response depending on how pre-cooked the value is already. I can try to make it clearer, but I doubt it can be much more concise. I will split out in separate method.

File: packages/serverpod/lib/src/server/server.dart (hunk 9 above)
Comment (nielsenko) [reply packages/serverpod/lib/src/server/server.dart]: Yes. It is actually removed in another PR, but I can do it in this instead/as well.

File: packages/serverpod/lib/src/server/server.dart (hunk 4 above)
Comment (nielsenko) [reply packages/serverpod/lib/src/server/server.dart]: Not sure I follow. The previous if is just what header line to put in the body. Can you elaborate?

Comment (nielsenko) [reply packages/serverpod/lib/src/server/server.dart]: Now I get it. Old code, but will refactor.

File: packages/serverpod/lib/src/server/server.dart (hunk 8 above)
Comment (nielsenko) [reply packages/serverpod/lib/src/server/server.dart]: Fixed in 34a19c5f. Much shorter and clearer.

File: packages/serverpod/lib/src/server/server.dart (hunk 4 above)
Comment (nielsenko) [reply packages/serverpod/lib/src/server/server.dart]: Fixed in 34a19c5f. Returns early and build body from stream.

File: packages/serverpod/lib/src/server/server.dart (hunk 3 above)
Comment (nielsenko) [reply packages/serverpod/lib/src/server/server.dart]: Fixed in 34a19c5f

File: packages/serverpod/lib/src/server/server.dart (hunk 9 above)
Comment (nielsenko) [reply packages/serverpod/lib/src/server/server.dart]: Fixed in 34a19c5f

File: packages/serverpod/lib/src/server/server.dart (hunk 4 above)
Comment (SandPod) [reply packages/serverpod/lib/src/server/server.dart]: I'm not sure I agree that the format now is easer to read. I was more looking for a:

```dart
if (allOK) {
  return Response.ok(body: Body.fromString('OK'));
}
// else all the compiled stuff when we had an error.
```

Comment (nielsenko) [reply packages/serverpod/lib/src/server/server.dart]: But this would be semantically different. We return the actual metrics, also when `allOk` is `true`. 
Not sure it is important, but what you suggest is breaking.

File: packages/serverpod/lib/src/server/server.dart (hunk 6 above)
Comment (nielsenko) [reply packages/serverpod/lib/src/server/server.dart:282]: I'll split `handleUriCall` in two. We don't actually need to do much of it when serving `/serverpod_cloud_storage`. If my analysis is correct, then that is the only single segment "endpoint" we have.

File: packages/serverpod/lib/src/server/server.dart (hunk 8 above)
Comment (SandPod) [reply packages/serverpod/lib/src/server/server.dart]: Thanks 🙏

File: packages/serverpod/lib/src/server/server.dart (hunk 4 above)
Comment (nielsenko) [reply packages/serverpod/lib/src/server/server.dart]: Sorry, my bad. You are right. Fixed in edc12d3c

**Review by SandPod — APPROVED** [review]
👍 LGTM!
